### PR TITLE
KAN-51 - instant deps introduced

### PR DESCRIPTION
**Pull request by DenShlk**
Changes Pool interface to work with transactions
Adds Instant implementation of deps, it puts resolved deps in pull, from where they can be processd immidiately.
Previous implementation is preserved and named Batched.

TODO:

- [x] implement new pool
- [x] fix incompatibility with the previous usages
- [x] run tests for both Batched and Instant deps

### CODE REVIEW

File: impl/deps/postgres/instant/instant.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+		ID:       id,
+		Resolved: false,
+	}, nil
+}
+
+// MakeGroup creates a new dependency group
+//
+// NOTE: the new group's id is generated via a UUID algorithm
+func (d *Deps) MakeGroup(ctx context.Context, ids ...string) (string, error) {
+	id := uuid.New().String()
```
Comment (ischenkx) [inline impl/deps/postgres/instant/instant.go:116]: Let's consider the following scenario:
1. User creates N dependencies
2. Resolves them all
3. Creates a group consisting of any subset of those dependencies

What happens to that group? Seems like it will never be resolved...

File: impl/deps/postgres/instant/instant.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	// dep hasn't been resolved, so we need to update groups
+	// we also return groups which have no unresolved deps
+	sql = `
+			UPDATE InstantGroups g
+			SET pending = pending - 1
+			FROM InstantGroupDependencies gd
+			WHERE gd.dependency_id = $1 AND gd.group_id = g.id
+			RETURNING (SELECT g.id WHERE g.pending = 0)` // returns new value
+	resolved, err := tx.Query(ctx, sql, dep)
+	var groups []string
```
Comment (ischenkx) [inline impl/deps/postgres/instant/instant.go:192]: Despite that `resolved` is automatically closed after a successful scanning in the loop below, we have to `defer resolved.Close()` in order to handle a case when an error happens while scanning.
https://pkg.go.dev/github.com/jackc/pgx#Rows

File: impl/deps/postgres/instant/instant.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+	// dep hasn't been resolved, so we need to update groups
+	// we also return groups which have no unresolved deps
+	sql = `
+			UPDATE InstantGroups g
+			SET pending = pending - 1
+			FROM InstantGroupDependencies gd
+			WHERE gd.dependency_id = $1 AND gd.group_id = g.id
+			RETURNING (SELECT g.id WHERE g.pending = 0)` // returns new value
+	resolved, err := tx.Query(ctx, sql, dep)
```
Comment (ischenkx) [inline impl/deps/postgres/instant/instant.go:191]: please handle an error (it might the cause of unexpected behaviour in the loop)

File: impl/deps/postgres/instant/instant.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+		var id *string
+		if err := resolved.Scan(&id); err != nil {
+			return fmt.Errorf("failed to scan the id of a group (please report if it ever happens): %s", err)
+		}
+		if id != nil {
+			groups = append(groups, *id)
+		}
+	}
+	// now we just add groups to their queue
+	if err := d.resolvedGroups.Write(ctx, groups...); err != nil {
```
Comment (ischenkx) [inline impl/deps/postgres/instant/instant.go:204]: it might be better to add an error message just like in line 209

File: impl/deps/postgres/instant/instant.go (hunk 2 above)
Comment (ischenkx) [reply impl/deps/postgres/instant/instant.go:192]: please check if there're any other similar mistakes

File: impl/deps/postgres/instant/instant.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+		return fmt.Errorf("failed to commit the transaction: %s", err)
+	}
+	return nil
+}
+
+func (d *Deps) Ready(ctx context.Context) (<-chan transactional.Object[string], error) {
+	return d.resolvedGroups.Read(ctx)
+}
+
+func (d *Deps) Run(_ context.Context) {}
```
Comment (ischenkx) [inline impl/deps/postgres/instant/instant.go:218]: why do we need a `Run` function? (it's not in the interface)
more importantly why do we need an unused context argument in a non-interface function?

File: testing/common/data/pool/pool_test.go [hunk 6]
```diff
@@ -4,7 +4,6 @@ import (
 	"context"
```
Comment (ischenkx) [inline testing/common/data/pool/pool_test.go:4]: please add a test case for a corner case scenario described in one of the comments above

**Review by ischenkx — CHANGES_REQUESTED** [review]

**Comment by ischenkx** [comment]
Well done @DenShlk !

File: impl/deps/postgres/instant/instant.go (hunk 5 above)
Comment (DenShlk) [reply impl/deps/postgres/instant/instant.go:218]: there is an interface in postgres/common.go which allows you to run postgres deps of both types

**Review by ischenkx — APPROVED** [review]
